tools/meteor-services/deploy.js

```diff
@@ -321,6 +334,122 @@ function canonicalizeSite(site) {
+  // `currentMessage` tracks what the current status message is for this version
+  initialize() {
+    this.start = new Date();
+    this.currentMessage = '';
+  }
+}
+
+// Poll the "version-status" endpoint for the build and deploy status
+// of a specified version ID with a polling configuration.
+// This will only end successfully when the polling endpoint reports that
+// the version deployment is finished. The version-status endpoints will report
+// messages pertaining to the status of the version, which will then be reported
+// directly to the user. When the poll is complete, it will return an object
+// with information about the final state of the version and the app.
+function pollForDeploy(pollingState, versionId, firstRun = true) {
+  const {
+    timeoutMs,
+    initialWaitTimeMs,
+    pollIntervalMs,
+    start,
+    currentMessage,
+  } = pollingState;
+  // If this is the first poll, we will wait for initialWaitTimeMs milliseconds 
+  // before beginning. We also want to initialize all of our tracking variables
+  if(firstRun) {
+    // Initialize the state of the poll, then begin polling after a pause
+    pollingState.initialize();
+    sleepMs(initialWaitTimeMs);
+    return pollForDeploy(pollingState, versionId, false);
+  }
+
+  // Do a call to the version-status endpoint for the specified versionId
+  const versionStatusResult = deployRpc({
+    method: 'GET',
+    operation: 'version-status',
+    site: versionId,
+    expectPayload: [],
+    printDeployURL: false,
+  });
+  
+  //Check the details of the Version Status response and compare message to last call
+  if(versionStatusResult &&
+    versionStatusResult.payload &&
+    versionStatusResult.payload.message) {
+      const message = versionStatusResult.payload.message;
+      if ( currentMessage !== message ) {
+        Console.info(message);
+        pollingState.currentMessage = message;
+      }
+  } else {
+    // If we did not get a valid Version Status response, just fail silently and
+    // keep polling as per usual – this may have just been a whiff from Galaxy
+    // We do the retry here because we might hit an error if we try to parse the
+    // result of the version-status call below.
+    if(new Date().getTime - start < timeoutMs) {
+      sleepMs(pollIntervalMs);
+      return pollForDeploy(pollingState, versionid, false);
+    }
+  }
+
+  const elapsed = new Date().getTime() - start;
+  const finishStatus = versionStatusResult.payload.finishStatus;
+  // Poll again if version isn't finished and we haven't exceeded the timeout
+  if(elapsed < timeoutMs && !finishStatus.isFinished) {
+    // Wait for a set interval and then poll again
+    sleepMs(pollIntervalMs);
+    return pollForDeploy(pollingState, versionId, false);
+  } else if (!finishStatus.isFinished) {
+    Console.info(`Polling timed out. To check the status of your app, visit
+    ${versionStatusResult.payload.galaxyUrl}. To change the default polling
```

Comment:
maybe we should just remember the URL from deploy instead of needing it to be here too?

Reply:
I personally find that more annoying because then you need to pass it into pollForDeploy (like 4 times) directly. Also I think it's worthwhile that the version-status endpoint give you everything you need, even if you might be able to get some of it elsewhere.